storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -492,15 +524,88 @@ func (s *commitLogSource) ReadIndex(
 	shardsTimeRanges result.ShardTimeRanges,
 	opts bootstrap.RunOptions,
 ) (result.IndexBootstrapResult, error) {
-	// FOLLOWUP(r): implement the commit log source returning
+	// FOLLOWUP(rartoul): implement the commit log source returning
 	// not indexed series metadata for the time range required.
 	// Try to cache some data on the commit log source itself
 	// from work done in ReadData(...) to help avoid rereading as
 	// much as possible.
 	// Also: it's now possible to cache the metadata and data
 	// for all namespaces in the first call to ReadData(...) since
 	// the source is used across all namespaces and then discarded after.
-	return result.NewIndexBootstrapResult(), nil
+	// Finally, we can also probably improve performance significantly by
+	// implementing a new interface in the Commitlog Reader/Iterator that
+	// only returns a given series once per file, although there is some
+	// trickiness there involving bufferpast and buffer future values.
+	if shardsTimeRanges.IsEmpty() {
+		return result.NewIndexBootstrapResult(), nil
+	}
+
+	// Setup predicates for skipping files / series at iterator and reader level.
+	readCommitLogPredicate := newReadCommitLogPredicate(
+		ns, shardsTimeRanges, s.opts, s.inspection)
+	readSeriesPredicate := newReadSeriesPredicate(ns)
+	iterOpts := commitlog.IteratorOpts{
+		CommitLogOptions:      s.opts.CommitLogOptions(),
+		FileFilterPredicate:   readCommitLogPredicate,
+		SeriesFilterPredicate: readSeriesPredicate,
+	}
+
+	// Create the commitlog iterator
+	iter, err := s.newIteratorFn(iterOpts)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create commit log iterator: %v", err)
+	}
+	defer iter.Close()
+
+	highestShard := s.findHighestShard(shardsTimeRanges)
+	// +1 so we can use the shard number as an index throughout without constantly
+	// remembering to subtract 1 to convert to zero-based indexing.
+	numShards := highestShard + 1
+	// Convert the map to a slice for faster lookups
+	bootstrapRangesByShard := make([]xtime.Ranges, numShards)
+	for shard, ranges := range shardsTimeRanges {
+		bootstrapRangesByShard[shard] = ranges
+	}
+
+	var (
+		indexResult    = result.NewIndexBootstrapResult()
+		indexResults   = indexResult.IndexResults()
+		indexBlockSize = ns.Options().IndexOptions().BlockSize()
+	)
+
+	for iter.Next() {
+		series, dp, _, _ := iter.Current()
+		if !s.shouldIncludeInIndex(
+			series.Shard, dp.Timestamp, highestShard, indexBlockSize, bootstrapRangesByShard) {
+			continue
+		}
+
+		segment, err := indexResults.GetOrAddSegment(
+			dp.Timestamp, ns.Options().IndexOptions(), s.opts.ResultOptions())
```

Comment:
Maybe pull these two vars so that these options aren't resolved by a method call each loop iter?

Reply:
yeah duh, my bad